File: go/odd_even_linked_list.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	even_dummy_head := &ListNode{0, nil}
	even_curr := even_dummy_head
	odd_dummy_head := &ListNode{0, nil}
	odd_curr := odd_dummy_head
	curr := head
	i := 1

	for curr != nil {
		if i%2 == 0 {
			even_curr.Next = curr
			even_curr = even_curr.Next
		} else {
			odd_curr.Next = curr
			odd_curr = odd_curr.Next
		}
		curr = curr.Next
		i++
	}

	odd_curr.Next = even_dummy_head.Next
	even_curr.Next = nil
	return odd_dummy_head.Next
}

File: python/odd-even-linked-list.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # Time complexity: O(n)
    # Space complexity: O(1)
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        i = 1
        odd_dummy_head = odd_curr = ListNode(0)
        even_dummy_head = even_curr = ListNode(0)
        curr = head
        
        while curr:
            if i % 2 == 0:
                even_curr.next = curr
                even_curr = even_curr.next
            else:
                odd_curr.next = curr
                odd_curr = odd_curr.next
            curr = curr.next
            i += 1
        
        odd_curr.next = even_dummy_head.next
        even_curr.next = None
        
        return odd_dummy_head.next